test_util/dcos_api_session.py

```diff
@@ -30,11 +30,111 @@ def get_args_from_env():
         'public_masters': os.environ['PUBLIC_MASTER_HOSTS'].split(','),
         'slaves': os.environ['SLAVE_HOSTS'].split(','),
         'public_slaves': os.environ['PUBLIC_SLAVE_HOSTS'].split(','),
-        'default_os_user': os.getenv('DCOS_DEFAULT_OS_USER', 'root'),
-        'ca_cert_path': os.getenv('DCOS_CA_CERT_PATH', None)}
+        'default_os_user': os.getenv('DCOS_DEFAULT_OS_USER', 'root')}
+
+
+class DcosUser:
+    """A lightweight user representation for grabbing the auth info and stashing it"""
+    def __init__(self, credentials: dict):
+        self.credentials = credentials
+        self.auth_token = None
+        self.auth_cookie = None
+        self.auth = None
+
+    def authenticate(self, dcos_url: str):
+        logging.info('Attempting authentication')
+        # explicitly use a session with no user authentication for requesting auth headers
+        r = requests.post(path_join(dcos_url, '/acs/api/v1/auth/login'), json=self.credentials, auth=None)
+        r.raise_for_status()
+        logging.info('Received authentication blob: {}'.format(r.json()))
+        self.auth_token = r.json()['token']
+        self.auth_cookie = r.cookies['dcos-acs-auth-cookie']
+        logging.info('Authentication successful')
+        # Set requests auth
+        self.auth = DcosAuth(self.auth_token)
+
+
+class DcosAuth(requests.auth.AuthBase):
+    def __init__(self, auth_token: str):
+        self.auth_token = auth_token
+
+    def __call__(self, request):
+        request.headers['Authorization'] = 'token={}'.format(self.auth_token)
+        return request
+
+
+class ARNodeApiClientMixin:
```

Comment:
>   the specific scenario of sending node argument to api_request in addition to host and port
That is not accurate. This is for when you are supplying node INSTEAD of host and port

> Can this be just extracted to a function call and api_request calls when the node argument (let's call it admin_router is not None, and we extract the updated host and port from it).

Once again, I don't think you read this right, the whole point of this extra node argument is that user's want to do something like this:
```
for node in cluster.all_slaves + cluster.masters:
    cluster.get('my/master/or/slave/endpoint', node=node)
```
those member variables are not provided with any knowledge of what port admin router is running on. If the user needs to supply the port then this whole thing is pointless

The reason this is a mixin is a is because we may want to create more specialized clients that need to know what a node is when spun off of a DcosApiSession.. For example, consider the test_util.marathon module, if we want to make a similar thing for health or logs (e.g. dump these logs when the test fails) then how do we teach that Api what a node is? Do we lock it into being a child of DcosApiSession? No, because it should be a sibling. Then we can do this:
```
class HealthClient(ApiClientSession): pass
# in reality there would be a lot of helpers and stuff instead of pass

# can be instantiated with only a url to health and not know anything about DC/OS
health = HealthClient(Url('http://dcos.dcos/system/health'))
```
```
from test_util.health import HealthClient

class HealthNodeApiClientSession(ARNodeApiClientMixin, HealthClient): pass
# Literally all the code you need to make the node-aware HealthClient

class DcosApiSession(ARNodeApiClientMixin, ApiClientSession):
     # ...

    @lazy_property
    def health(self):
         return HealthNodeApiClientSession(self.default_url.copy(path='system/health'))

dcos_api_session = DcosApiSession(*my_api_args)
dcos_api_session.health.get('some/endpoint', node=cluster.slave[0])
```
I hope this makes the value of the Mixin logic clear. A separate, independent health utility can be written free of all the junk in DcosApiSession and still get the benefits of knowing about how to make authenticated communication without knowing that agent admin router runs on different ports.

Furthermore, this sets up a way for anyone to easily add additional arguments or change a Api's signature without having the redefine and entire heirarchy of inter-dependent helper classes. You can imagein a MesosNodeClient mixin that automatically sets the correct port for master vs agent, or perhaps one could make a mixin that checks if a certain field is in the response body and then hard error if not, etc...

The great thing about mixins like that is you can arbitrarily combine as many as you want because of how python's super works in calling siblings before parents

> The assertions done in this Mixin really belong to either ApiClientSession or DcosApiSession. If point 1. is addressed, then this is automatically taken care.
They do not belong in ApiClientSession because ApiClientSession should have node idea what a 'node' is to stay generic. They do not belong in DcosApiSession because of the above reason we need to make this a mixin